File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visit = set()
        res = 0
        
        def dfs(r, c):
            if (
                r < 0 
                or c < 0
                or r == ROWS
                or c == COLS
                or (r, c) in visit
                or grid[r][c] == 0
            ):
                return 0
            visit.add((r, c))
            return (1 + dfs(r + 1, c) +
                        dfs(r, c + 1) + 
                        dfs(r - 1, c) + 
                        dfs(r, c - 1))
        
        for i in range(ROWS):
            for j in range(COLS):
                res = max(res, dfs(i, j))
        return res
            

File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1'){
                    if (dfs(i, j, grid)) {
                        count++;
                    }
                }
            }
        }
        return count;
    }
    private boolean dfs(int i, int j, char[][] grid) {
        if (i < 0 || j < 0 || i > grid.length - 1 || j > grid[0].length - 1 || grid[i][j] != '1')  
            return false;
        grid[i][j] = 0;
        dfs(i + 1, j, grid);
        dfs(i, j + 1, grid);
        dfs(i - 1, j, grid);
        dfs(i, j - 1, grid);
        return true;
    }
}
